iroh-p2p/src/node.rs

```diff
@@ -1095,7 +1209,288 @@ mod tests {
+            }
+            Some(n) => {
+                anyhow::bail!("unexpected network event: {:?}", n);
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Subscribed), received no event");
+            }
+        };
+
+        let peers = test_runner_a.client.gossipsub_all_peers().await?;
+        assert!(peers.len() == 1);
+        let got_peer = peers.get(0).unwrap();
+        assert_eq!(test_runner_b.peer_id, got_peer.0);
+        assert_eq!(&topic, got_peer.1.get(0).unwrap());
+
+        // get mesh peer for topic
+        let peers = test_runner_a
+            .client
+            .gossipsub_mesh_peers(topic.clone())
+            .await?;
+
+        assert!(peers.len() == 1);
+        assert_eq!(test_runner_b.peer_id, *peers.get(0).unwrap());
+
+        let peers = test_runner_a.client.gossipsub_all_mesh_peers().await?;
+        assert!(peers.len() == 1);
+        assert_eq!(&test_runner_b.peer_id, peers.get(0).unwrap());
+
+        let msg = Bytes::from(&b"hello world!"[..]);
+        test_runner_b
+            .client
+            .gossipsub_publish(topic.clone(), msg.clone())
+            .await?;
+
+        match test_runner_a.network_events.recv().await {
+            Some(NetworkEvent::Gossipsub(GossipsubEvent::Message { from, message, .. })) => {
+                assert_eq!(test_runner_b.peer_id, from);
+                assert_eq!(topic, message.topic);
+                assert_eq!(test_runner_b.peer_id, message.source.unwrap());
+                assert_eq!(msg.to_vec(), message.data);
+            }
+            Some(n) => {
+                anyhow::bail!("unexpected network event: {:?}", n);
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Message), received no event");
+            }
+        };
+
+        test_runner_b
+            .client
+            .gossipsub_unsubscribe(topic.clone())
+            .await?;
+        match test_runner_a.network_events.recv().await {
+            Some(NetworkEvent::Gossipsub(GossipsubEvent::Unsubscribed {
+                peer_id,
+                topic: unsubscribe_topic,
+            })) => {
+                assert_eq!(test_runner_b.peer_id, peer_id);
+                assert_eq!(topic, unsubscribe_topic);
+            }
+            Some(n) => {
+                anyhow::bail!("unexpected network event: {:?}", n);
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Unsubscribed), received no event");
+            }
+        };
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_dht() -> Result<()> {
+        // set up three nodes
+        // two connect to one
+        // try to connect via id
+        let addr_a: Multiaddr = "/ip4/127.0.0.1/tcp/5009".parse().unwrap();
+        let addr_b: Multiaddr = "/ip4/127.0.0.1/tcp/5010".parse().unwrap();
+        let addr_c: Multiaddr = "/ip4/127.0.0.1/tcp/5011".parse().unwrap();
```

Comment:
Would be really nice if there was a way not to hardcode ports but let the peers allocate their addresses dynamically so it works regardless of what else is using ports (maybe another run of these same tests!).